### PR TITLE
Query refactoring: split parse phase into fromXContent and toQuery for all queries

### PR DESCRIPTION
**Pull request by javanna**
This PR includes the work that was done in the query-refactoring branch in the last few months. All queries were refactored to split the current parse phase into tho phases:
1) `fromXContent` that returns an intermediate elasticsearch object which can be streamed over the wire
2) `toQuery` that converts the intermediate elasticsearch object into the corresponding lucene query

The first phase can be moved to the coordinating node, while the second phase will stay on the data node.

All queries have been refactored, plus lots of tests have been added for each query. Once this is merged we can start using this new infra and effectively move the parse phase to the coordinating node as the next step.

For all the detailed changes, see pull requests labelled with the [`:Query Refactoring` label](https://github.com/elastic/elasticsearch/pulls?q=is%3Aclosed+is%3Apr+label%3A%22%3AQuery+Refactoring%22).

Relates to #10217.

### CODE REVIEW

**Comment by s1monw** [comment]
LGTM LOL

File: core/src/main/java/org/apache/lucene/queries/ExtendedCommonTermsQuery.java [hunk 1]
```diff
@@ -76,6 +76,10 @@ public String getLowFreqMinimumNumberShouldMatchSpec() {
         return lowFreqMinNumShouldMatchSpec;
     }
 
+    public float getMaxTermFrequency() {
```
Comment (s1monw) [inline core/src/main/java/org/apache/lucene/queries/ExtendedCommonTermsQuery.java]: can we have javadocs here?

**Comment by javanna** [comment]
Marked this as breaking as it breaks backwards compatibility for plugins and the java api. Changes are included in the migrate guide. Also each PR that introduced breakage is [marked as breaking](https://github.com/elastic/elasticsearch/pulls?q=is%3Apr+label%3A%22%3AQuery+Refactoring%22+is%3Aclosed+label%3Abreaking).

**Comment by dadoonet** [comment]
@javanna do you plan to also change documentation at https://github.com/elastic/elasticsearch/tree/master/docs/java-api/query-dsl ?

**Comment by javanna** [comment]
yea we will need to do that, thanks for the reminder @dadoonet . opened #13800

**Pull request merged**